src/main/java/edu/wisc/my/ltiproxy/service/LTILaunchServiceImpl.java

```diff
@@ -21,38 +20,104 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 @Service
 public class LTILaunchServiceImpl implements LTILaunchService{
-    
+    protected final Logger logger = LoggerFactory.getLogger(getClass());
     private LTILaunchPropertyFileDao LTILaunchPropertyFileDao;
     
+    private static final String UTF8 = StandardCharsets.UTF_8.name();
+    
     @Autowired
     void setLTILaunchProperityFileDao(LTILaunchPropertyFileDao LTILaunchPropertyFileDao){
         this.LTILaunchPropertyFileDao = LTILaunchPropertyFileDao;
     }
 
     @Override
-    public JSONObject getFormData(String key, HttpServletRequest request) throws
-        JsonParseException, JsonMappingException, IOException, LtiSigningException, JSONException {
+    public URI getRedirectUri (String key, Map<String, String> headers) throws
+        JsonParseException, JsonMappingException, IOException, LtiSigningException, URISyntaxException {
+        URI result = null;
         
-        Map<String, String> launchParameters = LTILaunchPropertyFileDao.getLaunchParameters(key);
+        Map<String, String> prepParams = prepareParameters(key, headers);
+        LTIParameters ltiParams = signParameters(key, prepParams);
         
-        TreeMap<String, String> paramsToSign = new TreeMap<String, String>();
-        paramsToSign.putAll(launchParameters);
-        paramsToSign.putAll(getHeaders(key, request));
+        String actionUrl = ltiParams.getActionURL();
+        List<NameValuePair> formBody = buildFormBody(ltiParams.getSignedParameters());
         
-        LtiSigner ltiSigner = new LtiOauthSigner();
-        String ltiKey = LTILaunchPropertyFileDao.getProperty(key, "key");
-        String ltiSecret = LTILaunchPropertyFileDao.getProperty(key, "secret");
-        String ltiActionUrl = LTILaunchPropertyFileDao.getProperty(key, "actionURL");
-        Map<String, String> signedParameters = ltiSigner.signParameters(
-                paramsToSign, ltiKey, ltiSecret, ltiActionUrl, "POST");
+        CloseableHttpClient httpClient = null;
+        CloseableHttpResponse httpResp = null;
+
+        try {
+            httpClient = HttpClients.createDefault();
+            HttpPost httpPost = new HttpPost(actionUrl);
+            UrlEncodedFormEntity ent = new UrlEncodedFormEntity(formBody);
+            httpPost.setEntity(ent);
+
+            httpResp = httpClient.execute(httpPost);
+            
+            int status = httpResp.getStatusLine().getStatusCode();
+            logger.debug("status " + status);
+            for (Header header : httpResp.getAllHeaders()) {
+                logger.trace(header.getName() + " : " + header.getValue());
+            }
+            
+            HttpEntity respEntity = httpResp.getEntity();
+            byte[] b = EntityUtils.toByteArray(respEntity);
+            EntityUtils.consumeQuietly(respEntity);
+            
+            logger.trace(new String(b));
+            
+            if (303 == status) {
```

Comment:
magic number.  This possibly should be opened up as well to the 300-series, but right now I know that the service returns a 303 on success.